### PR TITLE
sign container image

### PR DESCRIPTION
**Pull request by tuananh**
Signed-off-by: Tuan Anh Tran <me@tuananh.org>

**1. Issue, if available:**

#921 

**2. Description of changes:**

this is wip. wanna get your thought on this before proceeding.

i added an entry in Makefile to allow `make release-test`: what it does is 
- do `publish-test` which is basically `publish` for current platform only to save time testing
- and then `sign-container-test` which will use cosign to sign controller and webhook image. use `cosign generate-key-pair` to generate a key pair for testing.
- user can then verify the container image with the pub key we provide with cli or admission controller or any of the policy engine management like opa or kyverno.
- cosign support aws kms so we can store our private key there. the [keyless feature](https://github.com/sigstore/cosign/blob/main/KEYLESS.md) of cosign is still experimental. we can consider using that once it's stable.

```
cosign verify --key cosign.pub public.ecr.aws/o9z7f6t4/controller:v0.5.0-35-gc3b37d9@sha256:a7d6d511505e6d5cfd5cb2b0f75258454238802e32bb54dbccf897a59575adb9 | jq

Verification for public.ecr.aws/o9z7f6t4/controller@sha256:a7d6d511505e6d5cfd5cb2b0f75258454238802e32bb54dbccf897a59575adb9 --
The following checks were performed on each of these signatures:
  - The cosign claims were validated
  - The signatures were verified against the specified public key
  - Any certificates were verified against the Fulcio roots.
[
  {
    "critical": {
      "identity": {
        "docker-reference": "public.ecr.aws/o9z7f6t4/controller"
      },
      "image": {
        "docker-manifest-digest": "sha256:a7d6d511505e6d5cfd5cb2b0f75258454238802e32bb54dbccf897a59575adb9"
      },
      "type": "cosign container image signature"
    },
    "optional": {
      "GIT_HASH": "c3b37d99bdfa84ed5fc28c544d5f02d640b3f70c",
      "GIT_VERSION": "v0.5.0-35-gc3b37d9-dirty"
    }
  }
```

**3. Does this change impact docs?**
- [ ] Yes, PR includes docs updates
- [ ] Yes, issue opened: *link to issue*
- [ ] No

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by tuananh** [comment]
@ellistarn can you take a look? do you want me to remove the `releast-test` and `publish-test` script in makefile?

File: Makefile [hunk 1]
```diff
@@ -87,4 +108,8 @@ website: ## Generate Docs Website
 toolchain: ## Install developer toolchain
 	./hack/toolchain.sh
 
+sign-container: ## cosign generate-key-pair --kms awskms:///alias/<myalias> to create a key pair
```
Comment (ellistarn) [inline Makefile:111]: Optional -- maybe simpler to just include this in publish.

Comment (ellistarn) [inline Makefile:111]: I think I need to check this out and play with it.

**Comment by ellistarn** [comment]
> @ellistarn can you take a look? do you want me to remove the `releast-test` and `publish-test` script in makefile?

I'd leave the `-test` versions out. Any manual testing/exploration will require mutating the Makefile beyond what these targets do, so I think it's a bit cleaner to keep the explicit targets that are used as part of some well known process.

**Comment by ellistarn** [comment]
> cosign support aws kms so we can store our private key there. the keyless feature of cosign is still experimental. we can consider using that once it's stable.
I'd love to jump straight to keyless to simplify things. Have you tried it? Of course, we can use KMS if necessary.

**Comment by tuananh** [comment]
> cosign support aws kms so we can store our private key there. the keyless feature of cosign is still experimental. we can consider using that once it's stable.
> I'd love to jump straight to keyless to simplify things. Have you tried it? Of course, we can use KMS if necessary.

i create an ECR repo and kms key to test and it works :)

**Comment by tuananh** [comment]
@ellistarn i change to use keyless sign of cosign and it works fine too. when publish, you will have to authen with an oidc (github/microsoft, etc...)

image

**Comment by ellistarn** [comment]
> @ellistarn i change to use keyless sign of cosign and it works fine too. when publish, you will have to authen with an oidc (github/microsoft, etc...)
> 
> image
> 
> image

Awesome! Github sounds perfect :)

File: Makefile [hunk 2]
```diff
@@ -76,6 +89,11 @@ publish: ## Generate release manifests and publish a versioned container image.
 	yq e -i ".controller.image = \"$$($(WITH_RELEASE_REPO) $(WITH_GOFLAGS) ko publish -B -t $(RELEASE_VERSION) $(RELEASE_PLATFORM) ./cmd/controller)\"" charts/karpenter/values.yaml
 	yq e -i ".webhook.image = \"$$($(WITH_RELEASE_REPO) $(WITH_GOFLAGS) ko publish -B -t $(RELEASE_VERSION) $(RELEASE_PLATFORM) ./cmd/webhook)\"" charts/karpenter/values.yaml
 	yq e -i '.version = "$(subst v,,${RELEASE_VERSION})"' charts/karpenter/Chart.yaml
+	make sign-container
```
Comment (ellistarn) [inline Makefile:92]:
```suggestion
	COSIGN_EXPERIMENTAL=1 cosign sign ${COSIGN_SIGN_FLAGS} ${RELEASE_REPO}/controller:${RELEASE_VERSION}
	COSIGN_EXPERIMENTAL=1 cosign sign ${COSIGN_SIGN_FLAGS} ${RELEASE_REPO}/webhook:${RELEASE_VERSION}
```

**Comment by ellistarn** [comment]
Hey @tuananh, it looks like the CI is failing due to the go package changes. Can you run `make verify` or manually run `go mod tidy` to make sure everything is cleaned up?

**Comment by tuananh** [comment]
@ellistarn i tried both but no changes to any file. 

i'm guessing it's because of this command in `make toolchain` always install the latest version?

```sh
GO111MODULE=on cat tools.go | grep _ | awk -F'"' '{print $2}' | xargs -tI % go install %
```

**Comment by ellistarn** [comment]
You should be able to pin any version you want in the `required` section of `tools/mod.go`. `go mod tidy` will clean everything up from there but won't change your requirements. If you're stuck, I can try to help make it compile.

**Comment by tuananh** [comment]
> You should be able to pin any version you want in the `required` section of `tools/mod.go`. `go mod tidy` will clean everything up from there but won't change your requirements. If you're stuck, I can try to help make it compile.

yes please

**Comment by ellistarn** [comment]
I'm finding myself in dependency hell. Minimally, I was able to install it manually -- perhaps we can't add this to tools yet, and the releaser will need to self install.

I'm also seeing this issue
```
➜  karpenter git:(sign-container-image) ✗ make publish-test
Login Succeeded
COSIGN_EXPERIMENTAL=1 cosign sign -a GIT_HASH=fbc898b548f5621a087af5e76a364615d3e8a3b1 -a GIT_VERSION=v0.5.1-9-gfbc898b-dirty -a BUILD_DATE='2021-12-07T06:55:21Z' public.ecr.aws/karpenter/controller:v0.5.1-9-gfbc898b
Generating ephemeral keys...
Retrieving signed certificate...
Your browser will now be opened to:
https://oauth2.sigstore.dev/auth/auth?access_type=online&client_id=sigstore&code_challenge=8m7lg9VhEUtqyE0qtzboCxnUZO-ieE1Gaz7ndIQg3Lo&code_challenge_method=S256&nonce=221mKEi1KO8tDGBmA7bQrvrX2gm&redirect_uri=http%3A%2F%2Flocalhost%3A63167%2Fauth%2Fcallback&response_type=code&scope=openid+email&state=221mKDgR1bALCekk7zFno7HOexa
Successfully verified SCT...
Error: signing [public.ecr.aws/karpenter/controller:v0.5.1-9-gfbc898b]: accessing entity: GET https://public.ecr.aws/v2/karpenter/controller/manifests/v0.5.1-9-gfbc898b: MANIFEST_UNKNOWN: Requested image not found
main.go:46: error during command execution: signing [public.ecr.aws/karpenter/controller:v0.5.1-9-gfbc898b]: accessing entity: GET https://public.ecr.aws/v2/karpenter/controller/manifests/v0.5.1-9-gfbc898b: MANIFEST_UNKNOWN: Requested image not found
make: *** [publish-test] Error 1
```

**Comment by ellistarn** [comment]
Got this working on my local repo, but the build is still broken due to dependencies 
```
RELEASE_REPO=$KO_DOCKER_REPO make publish-test
yq e -i ".controller.image = \"$(KO_DOCKER_REPO=767520670908.dkr.ecr.us-west-2.amazonaws.com/karpenter GOFLAGS="-tags=aws "-ldflags=-X=github.com/aws/karpenter/pkg/utils/project.Version=v0.5.1-9-gfbc898b"" ko publish -B -t v0.5.1-9-gfbc898b --platform=linux/amd64,linux/arm64 ./cmd/controller)\"" charts/karpenter/values.yaml
2021/12/08 17:48:44 Using base gcr.io/distroless/static:nonroot for github.com/aws/karpenter/cmd/controller
2021/12/08 17:48:45 Building github.com/aws/karpenter/cmd/controller for linux/amd64
2021/12/08 17:48:54 Building github.com/aws/karpenter/cmd/controller for linux/arm64
2021/12/08 17:49:58 Publishing 767520670908.dkr.ecr.us-west-2.amazonaws.com/karpenter/controller:v0.5.1-9-gfbc898b
2021/12/08 17:49:58 existing blob: sha256:e8614d09b7bebabd9d8a450f44e88a8807c98a438a2ddd63146865286b132d1b
2021/12/08 17:49:59 pushed blob: sha256:4c58f517fc24a42a9820fbf9ef90de5d8ed68bdfc3567c2d0ee15c7b5c5f61d3
2021/12/08 17:49:59 pushed blob: sha256:f57e1531375464f27342cc8aabb53e2f28f7dfad4a6b9ea21368ec7034cba6f2
2021/12/08 17:50:01 pushed blob: sha256:cf27b4fc0b7b105ef758a276883e372818566e7cfb21ec9ce874e62ff20c0f71
2021/12/08 17:50:01 767520670908.dkr.ecr.us-west-2.amazonaws.com/karpenter/controller@sha256:aadb4edbd6e8894cfa115b6ffe2efed7ef4febbdd802af3c698f6a030508272b: digest: sha256:aadb4edbd6e8894cfa115b6ffe2efed7ef4febbdd802af3c698f6a030508272b size: 753
2021/12/08 17:50:01 existing blob: sha256:f57e1531375464f27342cc8aabb53e2f28f7dfad4a6b9ea21368ec7034cba6f2
2021/12/08 17:50:02 pushed blob: sha256:8923342aafaf3b8af972d36292188f95021955a2386952362b5a489664f1f10c
2021/12/08 17:50:03 pushed blob: sha256:859e03b7461b2a512159493ef1504d2859ed37c05ed1ef781ff98394ea4799b5
2021/12/08 17:50:04 pushed blob: sha256:244b4c36df382fa3875618e574490e43e245b3104c2cb739930c9887f3232971
2021/12/08 17:50:04 767520670908.dkr.ecr.us-west-2.amazonaws.com/karpenter/controller@sha256:0971ec688c762505d9a6766af1bea46e8e7b50029f02db466f83fa8897f3357c: digest: sha256:0971ec688c762505d9a6766af1bea46e8e7b50029f02db466f83fa8897f3357c size: 753
2021/12/08 17:50:04 767520670908.dkr.ecr.us-west-2.amazonaws.com/karpenter/controller:v0.5.1-9-gfbc898b: digest: sha256:6586f894a0a5905fd19474cb80fbadac6b2ccda4f41e34d5a692de36cba88353 size: 529
2021/12/08 17:50:04 Published 767520670908.dkr.ecr.us-west-2.amazonaws.com/karpenter/controller:v0.5.1-9-gfbc898b@sha256:6586f894a0a5905fd19474cb80fbadac6b2ccda4f41e34d5a692de36cba88353
yq e -i ".webhook.image = \"$(KO_DOCKER_REPO=767520670908.dkr.ecr.us-west-2.amazonaws.com/karpenter GOFLAGS="-tags=aws "-ldflags=-X=github.com/aws/karpenter/pkg/utils/project.Version=v0.5.1-9-gfbc898b"" ko publish -B -t v0.5.1-9-gfbc898b --platform=linux/amd64,linux/arm64 ./cmd/webhook)\"" charts/karpenter/values.yaml
2021/12/08 17:50:07 Using base gcr.io/distroless/static:nonroot for github.com/aws/karpenter/cmd/webhook
2021/12/08 17:50:08 Building github.com/aws/karpenter/cmd/webhook for linux/amd64
2021/12/08 17:50:19 Building github.com/aws/karpenter/cmd/webhook for linux/arm64
2021/12/08 17:50:31 Publishing 767520670908.dkr.ecr.us-west-2.amazonaws.com/karpenter/webhook:v0.5.1-9-gfbc898b
2021/12/08 17:50:31 existing blob: sha256:e8614d09b7bebabd9d8a450f44e88a8807c98a438a2ddd63146865286b132d1b
2021/12/08 17:50:32 pushed blob: sha256:0d5f2ca76d3ed3343dca470366e571848c6f75290b562c804c4a19003b577b35
2021/12/08 17:50:32 pushed blob: sha256:f57e1531375464f27342cc8aabb53e2f28f7dfad4a6b9ea21368ec7034cba6f2
2021/12/08 17:50:34 pushed blob: sha256:38bc807a63e5beab23530e4926b7390665be4cccab08ae5a02e76c881efd50c9
2021/12/08 17:50:34 767520670908.dkr.ecr.us-west-2.amazonaws.com/karpenter/webhook@sha256:8b762e91542f032071b79080a72daba3ac1856dde3940269f42673d291e28e15: digest: sha256:8b762e91542f032071b79080a72daba3ac1856dde3940269f42673d291e28e15 size: 753
2021/12/08 17:50:34 existing blob: sha256:f57e1531375464f27342cc8aabb53e2f28f7dfad4a6b9ea21368ec7034cba6f2
2021/12/08 17:50:35 pushed blob: sha256:afa9fcf942824df3501016da871f4f646aab4ade5e0cf2c893c03afdfe9531c0
2021/12/08 17:50:36 pushed blob: sha256:859e03b7461b2a512159493ef1504d2859ed37c05ed1ef781ff98394ea4799b5
2021/12/08 17:50:37 pushed blob: sha256:020ea573a5b1562950cf2e9bcf45b7fbef0526cd30227de0ee30fecde03c7869
2021/12/08 17:50:37 767520670908.dkr.ecr.us-west-2.amazonaws.com/karpenter/webhook@sha256:9807d33d72956305d1e9576b4b3b8e88e09e586fddf1edc72051a95f38598b13: digest: sha256:9807d33d72956305d1e9576b4b3b8e88e09e586fddf1edc72051a95f38598b13 size: 753
2021/12/08 17:50:37 767520670908.dkr.ecr.us-west-2.amazonaws.com/karpenter/webhook:v0.5.1-9-gfbc898b: digest: sha256:4b6354ca8015a4e8bc063abac50fe700390d97007092c5138ce57593f671e207 size: 529
2021/12/08 17:50:37 Published 767520670908.dkr.ecr.us-west-2.amazonaws.com/karpenter/webhook:v0.5.1-9-gfbc898b@sha256:4b6354ca8015a4e8bc063abac50fe700390d97007092c5138ce57593f671e207
COSIGN_EXPERIMENTAL=1 cosign sign -a GIT_HASH=fbc898b548f5621a087af5e76a364615d3e8a3b1 -a GIT_VERSION=v0.5.1-9-gfbc898b-dirty -a BUILD_DATE='2021-12-07T06:55:21Z' 767520670908.dkr.ecr.us-west-2.amazonaws.com/karpenter/controller:v0.5.1-9-gfbc898b
Generating ephemeral keys...
Retrieving signed certificate...
Your browser will now be opened to:
https://oauth2.sigstore.dev/auth/auth?access_type=online&client_id=sigstore&code_challenge=_74r7bUTv3XlCPzh36TklblMri-ahStLLpouUratZqc&code_challenge_method=S256&nonce=221oph5XY5SWbXWZPHEWCWd4m5Z&redirect_uri=http%3A%2F%2Flocalhost%3A63951%2Fauth%2Fcallback&response_type=code&scope=openid+email&state=221opgA4PMbbC6jhGPStQYE6osO
Successfully verified SCT...
"767520670908.dkr.ecr.us-west-2.amazonaws.com/karpenter/controller" appears to be a private repository, please confirm uploading to the transparency log at "https://rekor.sigstore.dev" [Y/N]: Y
tlog entry created with index: 926527
Pushing signature to: 767520670908.dkr.ecr.us-west-2.amazonaws.com/karpenter/controller
COSIGN_EXPERIMENTAL=1 cosign sign -a GIT_HASH=fbc898b548f5621a087af5e76a364615d3e8a3b1 -a GIT_VERSION=v0.5.1-9-gfbc898b-dirty -a BUILD_DATE='2021-12-07T06:55:21Z' 767520670908.dkr.ecr.us-west-2.amazonaws.com/karpenter/webhook:v0.5.1-9-gfbc898b
Generating ephemeral keys...
Retrieving signed certificate...
Your browser will now be opened to:
https://oauth2.sigstore.dev/auth/auth?access_type=online&client_id=sigstore&code_challenge=k3mAIWmPPTuSQEMAbUrfMeKQeethAlLkGmm0p7hC-Lw&code_challenge_method=S256&nonce=221ov8TrPZ0KkYco5nbRU7J7E9Z&redirect_uri=http%3A%2F%2Flocalhost%3A63985%2Fauth%2Fcallback&response_type=code&scope=openid+email&state=221ov7qWaItJYHUvkzQSYGu7kyB
Successfully verified SCT...
"767520670908.dkr.ecr.us-west-2.amazonaws.com/karpenter/webhook" appears to be a private repository, please confirm uploading to the transparency log at "https://rekor.sigstore.dev" [Y/N]: Y
tlog entry created with index: 926528
Pushing signature to: 767520670908.dkr.ecr.us-west-2.amazonaws.com/karpenter/webhook
```

**Comment by tuananh** [comment]
i guess we can leave the cosign out of tools for now and releaser needs to install it manually

**Review by ellistarn — APPROVED** [review]

**Pull request merged**